Form: Fix show variation picker if empty

enejb opened the pull request:
This PR tries to fix the state where a user is shown a form with just the submit button and it is not clear how to proceed next. 

This PR instead will show the variation picker instead of just the submit button. 



https://github.com/user-attachments/assets/39985e8c-7196-40f3-89d9-4c5e2c432fb4



## Proposed changes:

* If we detect that we are just showing the submit button in the inner block. We show the button inserter that takes up the whole width of the page. 
* When you insert a block using a form inserted the field shows up as you would expect it to. (Before the submit button)
Unless you intentionally moved the submit button to not be the last element. 

### Other information:

- [x] Have you written new tests for your changes, if applicable?
- [x] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [x] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion


## Does this pull request change what data or activity we track or use?
No

## Testing instructions:
* Go to the block editor. 
* Insert the form block. Or just insteard a email field for example. 
* Remove all the input fields (expect for the submit button) 
* You should now see button inserter. 
* add a field. notice that it shows up as expected. ( before the button field but in the last place) 
* Add another field. Notice that is shows up as expected. ( before the button field but in the last place) 
* Move the button to not be the last block. Notice that the field shows up after the button block.

aaronrobertshaw commented on projects/plugins/jetpack/changelog/fix-remove-form-if-empty:
```diff
@@ -0,0 +1,4 @@
+Significance: minor
+Type: bugfix
+
+Forms: Show the form variation picker if you only gave the submit button"
```
```suggestion
Forms: Show the form variation picker if you only gave the submit button.
```

Nit: typo

aaronrobertshaw commented on projects/packages/forms/changelog/fix-remove-form-if-empty:
```diff
@@ -0,0 +1,4 @@
+Significance: minor
+Type: changed
+
+Forms: Revert back to the variation picker if the form only has the submit button."
```
```suggestion
Forms: Revert back to the variation picker if the form only has the submit button.
```

Nit: typo

aaronrobertshaw reviewed (commented):
Nice fix :sparkles: 

:white_check_mark: Variation picker displays when form is reduced to only submit button
:white_check_mark: Variation picker isn't shown when reverting the form to non empty state via undo
:white_check_mark: Dragging field blocks into an otherwise empty form via the list view clears the variation picker as expected

Overall this tests well for me. 

There was one very minor rough edge I noticed which was the block outline includes some extraneous space at the top and bottom when the variation picker is shown for a form with only the submit button in it.

| Freshly Inserted Form | Empty Form (only submit button) |
|---|---|
| ![Screenshot_17-Mar_15-24-06_6077](https://github.com/user-attachments/assets/04bf648e-2bba-4664-bbc5-6c9f1e09db04) | ![Screenshot_17-Mar_15-22-33_1222](https://github.com/user-attachments/assets/1a7998dc-fb89-4f28-bd20-f2c33312746b) |

ilonagl commented:
Great work and fix, @enejb! Everything works as expected. 👍

Just a small visual note—we’re showing a blue line hover state for the form template. It’d be better to either remove it or tweak it to include both the icon and title (right now, it's just the icon). If we go with the latter, the same should apply to the newly added form variation picker to keep them consistent. 

<img width="520" alt="Screenshot 2025-03-17 at 18 02 31" src="https://github.com/user-attachments/assets/1a720a65-ae67-492d-bb9e-ef46f4eda791" />

ilonagl reviewed (dismissed):
Great work and fix, @enejb! Everything works as expected. 👍

Just a small visual note—we’re showing a blue line hover state for the form template. It’d be better to either remove it or tweak it to include both the icon and title (right now, it's just the icon). If we go with the latter, the same should apply to the newly added form variation picker to keep them consistent.

<img width="520" alt="Screenshot 2025-03-17 at 18 02 31" src="https://github.com/user-attachments/assets/bc337151-14f8-4750-9175-caf6a9ff5509" />

enejb commented:
@aaronrobertshaw Good catch on the form and the padding around it. 

I pushed a change that addresses that. At first I try to preserve the styles that are currently applied. For example if you added a specific padding to the form but in my testing the any custom padding gets overwritten with the one that comes from the variation anyways. 

THis keeps things similar. 

@ilonagl I can try to tackle the changes in a different PR. 

Currently how things work is very similar how core does things with variations. 

See this screenshot from the columns block. 
![Screenshot 2025-03-17 at 1 48 00 PM](https://github.com/user-attachments/assets/00c5bfe9-dab6-4ff3-9b50-bd3e58718f72)

aaronrobertshaw commented on projects/packages/forms/src/blocks/contact-form/edit.js:
```diff
@@ -149,6 +152,10 @@ function JetpackContactFormEdit( { name, attributes, setAttributes, clientId, cl
 			/>
 		);
+
+		// Remove padding from the empty form.
+		blockProps.style.paddingTop = 0;
+		blockProps.style.paddingBottom = 0;
```
What about left and right padding? If you are resetting padding for the placeholder, shouldn't you reset it all?

enejb replied:
We currently don't apply a default left right padding anymore. But it does make sense to add this as well.

aaronrobertshaw replied:
No but the user might apply some and these changes forcibly reset that user selection. So either you override the user's choices completely or you don't.

aaronrobertshaw reviewed (commented):
This is getting closer. Thanks for iterating!

It looks like the left/right padding resets were missed. 

Also, while re-testing I noticed that there's a rogue `margin-top: 16px` style on the `.jetpack-contact-form` element. Tracking that style, I think it was a regression introduced via [#37589](https://github.com/Automattic/jetpack/pull/37589/files#r2000104410).

Given we're fixing spacing styles here already, it might be good to tweak the selector for the linked style such that it only applies in the context of the AI assistant.

<img width="669" alt="Screenshot 2025-03-18 at 1 06 52 pm" src="https://github.com/user-attachments/assets/763b2398-dcf6-4455-992c-442622a1b9f6" />

enejb committed:
Forms: If we are only showing the the single button block. Revert back to the variation picker.

enejb committed:
Removes the .jetpack-contact-form margin-top padding

enejb commented:
@aaronrobertshaw  I removed the margin-top mostly because I have no idea how it helps. 
I didn't notice any difference when I was using the AI tool or not. 

cc @renatoagds can you help me understand if this margin top is still needed? How can we added it only when needed?

aaronrobertshaw commented on projects/plugins/jetpack/extensions/blocks/ai-assistant/editor.scss:
```diff
@@ -223,7 +223,3 @@
 	gap: 8px;
 }
-
-.jetpack-contact-form {
-	margin-top: 16px;
-}
```
Without understanding the context behind this style's addition. I think it could be heavy handed to blindly remove it. The same way it was a clumsy addition in the first place.

A safer bet would be to wrap it in `:where()`. That way it can still apply if there isn't already styles defined for the form.

renatoagds replied:
@aaronrobertshaw and @enejb, sorry for the delay. It's been a long time since I did this, and I don't remember exactly the reasons, but I just tested it without, and it's working fine. Let's move forward, and if we receive something, we can iterate later.

enejb replied:
Thanks for the feedback @renatoagds!. 
Lets merge it without the top margin.

ilonagl commented:
> @ilonagl I can try to tackle the changes in a different PR.

Currently how things work is very similar how core does things with variations.

See this screenshot from the columns block.

Oh, gotcha! Thanks for sharing. It can be addressed in a different PR for sure.

enejb committed:
fix missing onlySubmitBlock export

enejb committed:
remove the reset since we are no longer adding a padding by defefault

enejb committed:
Revert the margin change.

enejb committed:
Insert field block in before the button!

simison reviewed (approved):
I haven't tested or reviewed code but wanted to accepd based on design since I had feedback on that previously. 

Looks really great!

dhasilva reviewed (approved):
LGTM, nice update!
